File: Examples/NoChat-Example/NoChat-Example/TGSystemMessageCellLayout.m
//
//  TGSystemMessageCellLayout.m
//  NoChat-Example
//
//  Copyright (c) 2016-present, little2s.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

#import "TGSystemMessageCellLayout.h"
#import "NOCMessage.h"
#import "UIFont+NoChat.h"
#import "NSAttributedString+NoChat.h"

@implementation TGSystemMessageCellLayout {
    UIEdgeInsets _textLabelInsets;
}

- (instancetype)initWithChatItem:(id<NOCChatItem>)chatItem cellWidth:(CGFloat)width
{
    self = [super init];
    if (self) {
        _reuseIdentifier = @"TGSystemMessageCell";
        _chatItem = chatItem;
        _width = width;
        _textLabelInsets = UIEdgeInsetsMake(2, 7, 2, 5);
        [self setupBackgroundImage];
        [self setupAttributedText];
        [self calculateLayout];
    }
    return self;
}

- (void)setupBackgroundImage
{
    _backgroundImage = [TGSystemMessageCellLayout systemMessageBackground];
}

- (void)setupAttributedText
{
    NSString *text = self.message.text;
    NSAttributedString *one = [[NSAttributedString alloc] initWithString:text attributes:@{ NSFontAttributeName: [TGSystemMessageCellLayout textFont], NSForegroundColorAttributeName: [TGSystemMessageCellLayout textColor] }];
    _attributedText = one;
}

- (void)calculateLayout
{
    self.height = 0;
    self.backgroundImageViewFrame = CGRectZero;
    self.textLabelFrame = CGRectZero;
    
    if (self.attributedText.length == 0) {
        return;
    }
    
    CGSize limitSize = CGSizeMake(ceil(self.width * 0.75), CGFLOAT_MAX);
    CGSize textLabelSize = [self.attributedText noc_sizeThatFits:limitSize];
    
    CGFloat vPadding = 4;
    
    self.textLabelFrame = CGRectMake(self.width/2 - textLabelSize.width/2, vPadding, textLabelSize.width, textLabelSize.height);
    self.backgroundImageViewFrame = CGRectMake(self.textLabelFrame.origin.x - _textLabelInsets.left, self.textLabelFrame.origin.y - _textLabelInsets.top, self.textLabelFrame.size.width + _textLabelInsets.left + _textLabelInsets.right, self.textLabelFrame.size.height + _textLabelInsets.top + _textLabelInsets.bottom);
    
    self.height = vPadding * 2 + self.backgroundImageViewFrame.size.height;
}

- (NOCMessage *)message
{
    return (NOCMessage *)self.chatItem;
}

@end

@implementation TGSystemMessageCellLayout (TGStyle)

+ (UIFont *)textFont
{
    static UIFont *_textFont = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _textFont = [UIFont systemFontOfSize:13];
    });
    return _textFont;
}

+ (UIColor *)textColor
{
    return [UIColor whiteColor];
}

+ (UIColor *)textBackgroundColor
{
    static UIColor *_textBackgroundColor = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _textBackgroundColor = [UIColor colorWithWhite:0.2 alpha:0.25];
    });
    return _textBackgroundColor;
}

+ (UIImage *)systemMessageBackground
{
    static UIImage *_systemMessageBackground = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        UIImage *rawImage = [self generateSystemMessageBackground];
        _systemMessageBackground = [rawImage stretchableImageWithLeftCapWidth:(NSInteger)(rawImage.size.width/2) topCapHeight:(NSInteger)(rawImage.size.height/2)];
    });
    return _systemMessageBackground;
}

+ (UIImage *)generateSystemMessageBackground
{
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(20, 20), NO, 0);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    UIBezierPath *path = [UIBezierPath bezierPathWithRoundedRect:CGRectMake(0, 0, 20, 20) cornerRadius:10];
    
    UIColor *color = [self textBackgroundColor];
    [color setFill];
    
    CGContextAddPath(context, path.CGPath);
    CGContextFillPath(context);
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Examples/NoChat-Swift-Example/NoChat-Swift-Example/TGSystemMessageCellLayout.swift
//
//  TGSystemMessageCellLayout.swift
//  NoChat-Swift-Example
//
//  Copyright (c) 2016-present, little2s.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import NoChat

class TGSystemMessageCellLayout: NSObject, NOCChatItemCellLayout {
    
    var reuseIdentifier: String = "TGSystemMessageCell"
    var chatItem: NOCChatItem
    var width: CGFloat
    var height: CGFloat = 0
    
    var message: Message {
        return chatItem as! Message
    }
    
    var backgroundImageViewFrame = CGRect.zero
    var backgroundImage: UIImage?
    var textLabelFrame = CGRect.zero
    var attributedText: NSAttributedString?
    
    required init(chatItem: NOCChatItem, cellWidth width: CGFloat) {
        self.chatItem = chatItem
        self.width = width
        super.init()
        setupBackgroundImage()
        setupAttributedText()
        calculate()
    }
    
    func calculate() {
        height = 0
        backgroundImageViewFrame = CGRect.zero
        textLabelFrame = CGRect.zero
        
        guard let at = attributedText, at.length > 0 else {
            return
        }
        
        let limitSize = CGSize(width: ceil(width * 0.75), height: CGFloat.greatestFiniteMagnitude)
        let textLabelSize = at.noc_sizeThatFits(size: limitSize)
        
        let vPadding = CGFloat(4)
        
        let textLabelInsets = Style.textInsets
        
        textLabelFrame = CGRect(x: width/2 - textLabelSize.width/2, y: vPadding, width: textLabelSize.width, height: textLabelSize.height)
        
        backgroundImageViewFrame = CGRect(x: textLabelFrame.origin.x - textLabelInsets.left, y: textLabelFrame.origin.y - textLabelInsets.top, width: textLabelFrame.width + textLabelInsets.left + textLabelInsets.right, height: textLabelFrame.height + textLabelInsets.top + textLabelInsets.bottom)
        
        height = vPadding * 2 + backgroundImageViewFrame.height
    }
    
    private func setupBackgroundImage() {
        backgroundImage = Style.systemMessageBackground
    }
    
    private func setupAttributedText() {
        let text = message.text
        let one = NSAttributedString(string: text, attributes: convertToOptionalNSAttributedStringKeyDictionary([convertFromNSAttributedStringKey(NSAttributedString.Key.font): Style.textFont, convertFromNSAttributedStringKey(NSAttributedString.Key.foregroundColor): Style.textColor]))
        attributedText = one
    }
    
    struct Style {
        static let textInsets = UIEdgeInsets(top: 2, left: 7, bottom: 2, right: 5)
        static let textFont = UIFont.systemFont(ofSize: 13)
        static let textColor = UIColor.white
        static let textBackgroundColor = UIColor(white: 0.2, alpha: 0.25)
        
        static let systemMessageBackground: UIImage? = {
            if let rawImage = TGGenerateSystemMessageBackground() {
                let image = rawImage.stretchableImage(withLeftCapWidth: Int(rawImage.size.width/2), topCapHeight: Int(rawImage.size.height/2))
                return image
            } else {
                return nil
            }
        }()
    }
    
}

fileprivate func TGGenerateSystemMessageBackground() -> UIImage? {
    UIGraphicsBeginImageContextWithOptions(CGSize(width: 20, height: 20), false, 0)
    
    guard let context = UIGraphicsGetCurrentContext() else {
        return nil
    }

    let path = UIBezierPath(roundedRect: CGRect(x: 0, y: 0, width: 20, height: 20), cornerRadius: 10)
    
    let color = TGSystemMessageCellLayout.Style.textBackgroundColor
    color.setFill()
    
    context.addPath(path.cgPath)
    context.fillPath()
    
    let image = UIGraphicsGetImageFromCurrentImageContext()
    
    UIGraphicsEndImageContext()
    
    return image
}


// Helper function inserted by Swift 4.2 migrator.
fileprivate func convertToOptionalNSAttributedStringKeyDictionary(_ input: [String: Any]?) -> [NSAttributedString.Key: Any]? {
	guard let input = input else { return nil }
	return Dictionary(uniqueKeysWithValues: input.map { key, value in (NSAttributedString.Key(rawValue: key), value)})
}

// Helper function inserted by Swift 4.2 migrator.
fileprivate func convertFromNSAttributedStringKey(_ input: NSAttributedString.Key) -> String {
	return input.rawValue
}
